### DISCUSSION TITLE
Dynamic exiting of sampling but preserving the sampler(s)' state

### DISCUSSION THREAD

**Discussion by ben18785** [pints-team/pints#1637] [Ideas/Science]
I've been doing a fair bit of neural net training recently and one thing that's nice with some libraries is that they allow you to view training progress dynamically over time. On the basis of this, users can decide to halt training earlier than the max number of iterations. When they halt training, the neural net keeps its state. Users can then, later, if needs be, re-continue training.

My question is, could we do this in PINTS? I'm envisioning that we could have some sort of dynamic visualisation of sampling progress and a "halt early" button.

I know we can sort of do this with ask/tell but, to be honest, I imagine many users will use the controller objects as their first port of call. Giving them dynamic feedback on sampling progress and the option to halt sampling but maintain the internal state of the samplers could be really nice and would make inference a bit more pleasurable (I frequently start a sampler with too many iterations, for example, and only realise after a minute or so has elapsed).

**Comment by MichaelClerx** [comment]
See also https://github.com/pints-team/pints/issues/194, https://github.com/pints-team/pints/issues/224, and I'm sure we've talked about writing visualisation code using ask-and-tell several times

I'm not a fan of having this idea in the main controllers, as they're getting pretty complex already

**Comment by ben18785** [comment]
@MichaelClerx thanks. I'm glad we've been thinking similarly! Yes, I don't think it'd be a main controller function: rather, a separate function (or a method that can be called in the controller, perhaps?).

**Comment by MichaelClerx** [comment]
I'd be very tempted to slap the "student project" label on this! Sounds like a really fun/interesting thing a bsc student could do?

**Comment by MichaelClerx** [comment]
(Could lead into smarter criteria for automatic termination, I guess)

**Comment by ben18785** [comment]
Good idea! Have added it as a student project.
